Sources/SwiftKafka/KafkaConfig.swift

```diff
@@ -0,0 +1,112 @@
+ */
+public struct KafkaConfig {
+    private final class _Internal {
+        /// Pointer to the `rd_kafka_conf_t` object managed by `librdkafka`
+        private(set) var pointer: OpaquePointer
+
+        /// Initialize internal `KafkaConfig` object with default configuration
+        init() {
+            self.pointer = rd_kafka_conf_new()
+        }
+
+        /// Initialize internal `KafkaConfig` object through a given `rd_kafka_conf_t` pointer
+        init(pointer: OpaquePointer) {
+            self.pointer = pointer
+        }
+
+        deinit {
+            rd_kafka_conf_destroy(pointer)
+        }
+
+        func value(forKey key: String) -> String? {
+            let value = UnsafeMutablePointer<CChar>.allocate(capacity: KafkaClient.stringSize)
+            defer { value.deallocate() }
+
+            var valueSize = KafkaClient.stringSize
+            let configResult = rd_kafka_conf_get(
+                pointer,
+                key,
+                value,
+                &valueSize
+            )
+
+            if configResult == RD_KAFKA_CONF_OK {
+                return String(cString: value)
+            }
+            return nil
+        }
+
+        func set(_ value: String, forKey key: String) throws {
+            let errorChars = UnsafeMutablePointer<CChar>.allocate(capacity: KafkaClient.stringSize)
+            defer { errorChars.deallocate() }
+
+            let configResult = rd_kafka_conf_set(
+                pointer,
+                key,
+                value,
+                errorChars,
+                KafkaClient.stringSize
+            )
+
+            if configResult != RD_KAFKA_CONF_OK {
+                let errorString = String(cString: errorChars)
+                throw KafkaError(description: errorString)
+            }
+        }
+
+        func createDuplicate() -> _Internal {
+            let duplicatePointer: OpaquePointer = rd_kafka_conf_dup(self.pointer)
+            return .init(pointer: duplicatePointer)
+        }
+    }
+
+    private var _internal: _Internal
+
+    public init() {
+        self._internal = .init()
+    }
+
+    public var pointer: OpaquePointer {
+        return self._internal.pointer
+    }
+
+    /// Retrieve value of configuration property for `key`
+    public func value(forKey key: String) -> String? {
+        return self._internal.value(forKey: key)
+    }
+
+    /// Set configuration `value` for `key`
+    public mutating func set(_ value: String, forKey key: String) throws {
+        // Copy-on-write mechanism
+        if !isKnownUniquelyReferenced(&(self._internal)) {
+            self._internal = self._internal.createDuplicate()
+        }
+
+        try self._internal.set(value, forKey: key)
+    }
+}
+
+extension KafkaConfig: Equatable {
+    public static func == (lhs: KafkaConfig, rhs: KafkaConfig) -> Bool {
+        return lhs._internal === rhs._internal
+    }
+}
```

Comment:
Why isn't the default `Equatable` conformance working here? `OpaquePointer` conforms to `Equatable`. 

Also if we had `Equatable` we should probably just add `Hashable` as well.

Reply:
According to the build logs, `_Internal` should conform to `Equatable` if we want to use the default conformance. But then we would have to somehow make `_Internal` conform to `Equatable` as well — for me the default conformance did not work here either

Reply:
Can you please check again. This works:
```swift
struct Foo: Hashable {
    let pointer: OpaquePointer
}
```

Also can you conform to `Hashable` instead of `Equatable`